docs/spec/dataclasses.rst

```diff
@@ -526,6 +529,52 @@ This includes, but is not limited to, the following semantics:
   dataclass must be explicitly annotated as e.g. ``x: ClassVar[Final[int]] =
   3``.
 
+Converters
+^^^^^^^^^^
+
+The ``converter`` parameter can be specified in a field definition to provide a callable
+used to convert values when assigning to the associated attribute. This feature allows for automatic type
+conversion and validation during attribute assignment.
+
+Converter behavior:
+
+* The converter is used for all attribute assignment, including assignment
+  of default values, assignment in synthesized ``__init__`` methods
+  and direct attribute setting (e.g., ``obj.attr = value``).
+* The converter is not used when reading attributes, as the attributes should already have been converted.
+
+Typing rules for converters:
+
+* The ``converter`` must be a callable that must accept a single positional argument
+  (but may optionally accept other arguments, which are ignored for typing purposes).
+* The type of the first positional parameter provides the type of the synthesized ``__init__`` parameter
+  associated with the field.
+* The return type of the callable must be assignable to the field's declared type.
+* If ``default`` or ``default_factory`` are provided, the type of the default value should be
+  assignable to the first positional parameter of the ``converter``
+
+Example usage:
+
+.. code-block:: python
+
+    def str_or_none(x: Any) -> str | None:
+        return str(x) if x is not None else None
+
+    @custom_dataclass
+    class Example:
+        int_field: int = field(converter=int)
+        str_field: str | None = field(converter=str_or_none)
```

Comment:
since `dataclasses.field` isn't going to gain the `converter=` kwarg: in your example can you show the definition of `field` as a `field_specifier` of `custom_dataclass` to show that this is not `dataclasses.field`?

Reply:
How about `custom_field`?